### PR TITLE
GoNextToSubgoal to replace GoToObj

### PR DESCRIPTION
**Pull request by rizar**
This was surprisingly easy. I just added the logic for looking for the object to `GoToObj`, and it worked immediately. I tested on 200 levels:

**Master:**
```
(babyai) ╭─dzmitry@dzmitry-ThinkPad-X1-Carbon-5th ~/Dist/baby-ai-game  ‹dima-byebye-gotoobj*› 
╰─$ python scripts/eval_bot.py  --num_runs=200 --level=BossLevel
       BossLevel: 100.0%, r=0.945, s=86.47
total time: 61.2s
4387296
```

**PR**
```
(babyai) ╭─dzmitry@dzmitry-ThinkPad-X1-Carbon-5th ~/Dist/baby-ai-game  ‹dima-byebye-gotoobj*› 
╰─$ python scripts/eval_bot.py  --num_runs=200 --level=BossLevel
       BossLevel: 100.0%, r=0.945, s=86.39
total time: 64.6s
5193343
```

In terms of wall time, the bot becomes somewhat slower, but I am hopeful to mitigate this by adding a cache to `shortest_path`, because that function is often repeatedly called with the same arguments.

### CODE REVIEW

**Comment by rizar** [comment]
I have switched to using tuple comparison instead of `numpy.array.equal` (see 7073870) and saved 30% of wall time:

```
╰─$ python scripts/eval_bot.py  --num_runs=200 --level=BossLevel
       BossLevel: 100.0%, r=0.945, s=86.39
total time: 37.6s
total episode_steps: 17277
total bfs: 60248
total bfs steps: 5193343
```

File: babyai/bot.py [hunk 1]
```diff
@@ -317,154 +317,6 @@ def take_action(self, action):
         return True
 
 
-class GoToObjSubgoal(Subgoal):
-    def __init__(self, bot=None, datum=None, adjacent=False):
-        '''If adjacent=True, then the bot would only go two cells away from the objective (with free cell in between)'''
```
Comment (maximecb) [inline babyai/bot.py]: I think this comment should be rewritten/clarified. If adjacent is set, then the bot *will* go two cells away? What do you mean by "with free cell in between". It will leave a free cell in between it and the objective?

File: babyai/bot.py [hunk 2]
```diff
@@ -317,154 +317,6 @@ def take_action(self, action):
...
-                    # New subgoal: go next to the object
-                    self.bot.stack.append(GoNextToSubgoal(self.bot, obj_pos, reason='GoToObj', adjacent=self.adjacent))
-                    return True
-
-        # CASE 2: The bot hasn't seen the object. The action will count for the new subgoal
-        self.bot.stack.append(ExploreSubgoal(self.bot))
-        return False
-
-
 class GoNextToSubgoal(Subgoal):
```
Comment (maximecb) [inline babyai/bot.py]: Can you add a class-level comment explaining what this subgoal does in some detail?

**Comment by maximecb** [comment]
How does GoNextTo improve on GoToObj?

Can you test on more than 200 levels? That's not enough given how often we keep running into rare edge cases.

**Comment by rizar** [comment]
> How does GoNextTo improve on GoToObj?

It doesn't. The purpose of this PR is to simplify things and reduce code duplication.

> Can you test on more than 200 levels? That's not enough given how often we keep running into rare edge cases.

Agreed! The compute that I am normally using for running 1M demos is currently not feeling well. Will test this as soon as it is back up.

All you requests for comments will now be duly fulfilled.

File: babyai/bot.py (hunk 2 above)
Comment (rizar) [reply babyai/bot.py]: I wrote smth, let me know if it is understandable. I don't yet understand all of the flags though, for them I just copied the docstring that Salem had written.

File: scripts/eval_bot.py [hunk 3]
```diff
@@ -98,6 +98,8 @@
     num_success = 0
     total_reward = 0
     total_steps = []
+    total_bfs = 0
```
[the remark this answers is not in the document]
Comment (rizar) [reply scripts/eval_bot.py]: That's just me collecting some stats about the shortest path search. Very cheap to collect and very useful for diagnostics.

File: babyai/bot.py (hunk 2 above)
Comment (maximecb) [reply babyai/bot.py]: Looks good

**Comment by rizar** [comment]
This is now tested, doesn't see to introduce new bugs, old bugs are being fixed in #27
